File: java/permutationsequence.java
public class Solution {
    public String getPermutation(int n, int k) {
        // Start typing your Java solution below
        // DO NOT write main() function
        if(n==0) return "";
        if(n==1) return "1";
        int t = 1;
        int sn = n;
        k--;
        while(sn-1>0){
            t*=(sn-1);
            sn--;
        }
        boolean tt[] = new boolean[n];
        StringBuilder sb = new StringBuilder("");
        for(int i=0;i<n;i++){
            int r = k/t;
            k %= t;
            if(n-1-i!=0){
                t /= (n-1-i);
            }
            
            int j=0;
            int kk=0;
            while(j<r || tt[kk] == true ){
                if(tt[kk]==false){
                    j++;
                }
                kk++;
            }
            tt[kk] = true;
            sb.append((char)(kk+'1'));
            
        }
        return sb.toString();
    }
}

File: cpp/PermutationSequence.cpp
class Solution {
public:
    string getPermutation(int n, int k) {
        if(n==0) {
            return "";
        }
        string perm = "";
        int factorial = 1;
        k--;
        for(int i=1;i<n;++i) {
            factorial*= i;
        }
        vector<int> remainingNumbers(n);
        for(int i=0;i<n;++i) {
            remainingNumbers[i] = i+1;
        }
        for(int i=0;i<n;i++) {
            int index = k/factorial;
            k = k%factorial;
            if(factorial!=1) {
                factorial /= (n-1-i);
            }
            int j=0;
            while(true) {
                if(remainingNumbers[j]<0) {
                    j++;
                } else if(index==0) {
                    perm += char('0' + remainingNumbers[j]);
                    remainingNumbers[j] = -1;
                    break;
                } else {
                    --index;
                    j++;
                }
            }
        }
        return perm;
    }
};
